### DISCUSSION TITLE
AWS Returns 405 on Image Upload

### DISCUSSION THREAD

**Discussion by justinhandley** [outline/outline#1813] [General] [answered]
I have set up AWS, and I can connect using the credentials via other apps (like a 3rd party chrome extension will let me use the credentials to create folders and files in the s3 bucket)

However, when I try to upload via outline, I get a 405 response from Amazon, and then a 404 / not found on trying to display the image, so the image just shows as a broken image icon.

I'm getting the same behavior on a one-click Heroku install and on a local dev setup.  I can't find any documentation about the proper way to set up an S3 bucket to make this work - does that documentation exist?  Or does anyone have a working example they could share?

Justin

**Comment by tommoor** [comment]
405 is `Method Not Allowed`, which means your S3 bucket must not be setup to receive uploads?

**Comment by justinhandley** [comment]
So, as I said above, I can use the same credentials (literally copied from my .env) to connect via a chrome s3 browser extension and can add and delete both folders and images.  I have also confirmed with an S3 support representative that my user and s3 permissions are set up correctly.  Do you have a sample bucket permission policy?  Do you do anything exceptional?  If you can get me clarity on the setup that fixes this, I'm happy to do a pull request with documentation which is really a missing feature here.

**Comment by justinhandley** [comment]
@tommoor - this is the most recent update from Amazon on this issue.  From what they are saying, though, it is a core software issue, so I'm very surprised anyone is able to upload images if this is true.

Greetings from AWS! 

Thank you for the additional information. My name is Sristi and I will be assisting you with the case in the absence of my colleague Vaishakh, who is off-shift at the moment.

According to my understanding, I reviewed your configuration and these are my findings: 
---------------------------------------------------------------------------------------
Using Internal tools, I can see that the IAM user 'xxx' in AWS Account ID:xxxxx has Admin access which allows the user full access to resources in AWS. Further, I checked  S3 Bucket Configuration and see that the Bucket policy is granting access only to IAM user 'xxxx. Thus, I can confirm that the Bucket policy and IAM user is setup correctly from the permission perspective. Also, make sure to use credentials of IAM user 'xxxx' in your application to avoid permission related issues.

Generally, the error "MethodNotAllowed" indicates that the user agent has requested a valid resource using an invalid HTTP method, error usually happens if the HTTP method is not allowed or if the POST request is not properly done.[1]

Going through the case details and S3 logs from the request ID pair you have shared, I found out that your client is doing a "REST.POST.OBJECT" operation. Upon checking the headers I could see that your application was making POST requests to your S3 bucket:

Host: xxxxx.s3.amazonaws.com
Origin: https://xxxx.herokuapp.com

The POST operation adds an object to a specific bucket using HTML forms. POST is an alternative form of PUT that enables browser-based uploads as a way of putting object in buckets. Parameters that are passed to PUT via HTTP Headers are instead passed as form fields to POST in the multipart/form-data encoded message body.[2]

Please note that, S3 only supports browser based POST uploads(making the post directly to the s3 bucket), wherein all the parameters are sent in the form data in the Body of the request and not in the Headers. So, the POST method is failing when request is made from your application.

However, You can upload files directly to Amazon S3 through a browser using HTTP POST requests. Amazon S3 support HTTP POST using Browser. Additionally, You can make PUT request instead of POST.

For more information on "Browser-Based Upload using HTTP POST ", Please refer document[3][4].

I hope you found the above information useful. Please feel free to get back to me in case you have any queries or concerns and I will be glad to assist you!

**Reply by tommoor** [reply]
This isn't super helpful, they're just describing how POST uploads work – could you try setting `AWS_S3_FORCE_PATH_STYLE=false` as an environment variable on the server if you haven't already?

It might also be useful to see what the error looks like in developer tools – you can black out sensitive pieces.

**Comment by justinhandley** [comment]
And one more, the links referenced above.

References:
[1] List of Error Codes: [https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#ErrorCodeList](https://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html#ErrorCodeList)
[2] POST Object: [https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html)
[3] Authenticating Requests in Browser-Based Uploads Using POST: [https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-UsingHTTPPOST.html](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-UsingHTTPPOST.html)
[4] Browser-Based Upload using HTTP POST: [https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-post-example.html)

**Comment by justinhandley** [comment]
I did set this to no effect

`AWS_S3_FORCE_PATH_STYLE=false`

Console:

<img width="1679" alt="Screen Shot 2021-01-22 at 8 29 39 PM" src="https://user-images.githubusercontent.com/9010117/105565290-805ca000-5cfc-11eb-9bb1-311378337057.png">

Network tab for the failed upload, with some random letters:

Request URL: https://pirateandfox-sops.s3.amazonaws.com/pirateandfox-sops
Request Method: POST
Status Code: 405 Method Not Allowed
Remote Address: 52.217.98.156:443
Referrer Policy: no-referrer
Response Headers
view source
Access-Control-Allow-Methods: HEAD, GET, PUT, POST, DELETE
Access-Control-Allow-Origin: *
Allow: HEAD, DELETE, GET, PUT
Connection: close
Content-Type: application/xml
Date: Sat, 23 Jan 2021 00:30:07 GMT
Server: AmazonS3
Transfer-Encoding: chunked
Vary: Origin, Access-Control-Request-Headers, Access-Control-Request-Method
x-amz-id-2: mAhPhosCMY8KthDf00lZi35CJRVk9KlOVaLKsEW8dsfsadfsdafs5pd3xWEws=
x-amz-request-id: EH9X5WCMDRASFK1W
Request Headers
view source
Accept: */*
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
Connection: keep-alive
Content-Length: 322938
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary9qIURkAgI2E73BgZ
Host: pirateandfox-sops.s3.amazonaws.com
Origin: https://sops.pirateandfox.com
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
User-Agent: Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.141 Mobile Safari/537.36
Form Data
view source
view decoded
Cache-Control: max-age=31557600
Content-Type: image/jpeg
acl: private
key: uploads/2ed64f25-8e42-420e-9683-aebd7f80eef6/cec8a9ba-13bf-42fe-8d25-9b80e80d39fe/Image-03.jpg
policy: eyJjb25kaXRpb25zIjpbeyJidWNrZXQiOiJwaXsdfsdfsadfdsIn1dLCJleHBpcmF0aW9uIjoiMjAyMS0wMS0yNFQwMDozMDowN1oifQ==
x-amz-algorithm: AWS4-HMAC-SHA256
x-amz-credential: AKfdsafdsaN/20210123/us-east-1/s3/aws4_request
x-amz-date: 20210123T003007Z
x-amz-signature: 5cb4331b9fdsafdsafds101853a9096a6ac5e199bdd7deb6c
file: (binary)

**Comment by ersel** [comment, marked as answer]
Could it be related to this? 
https://forums.aws.amazon.com/thread.jspa?threadID=297751

Request URL: https://pirateandfox-sops.s3.amazonaws.com**/pirateandfox-sops**

Adding bucket name in the request URL might be problematic with presigned POST URL...

**Reply by justinhandley** [reply]
So, I guess it could be, but that is happening in the software as far as I can tell, and if the software works for anyone, it should be universal I think?

So, in the .env variables there are two different ones..

AWS_S3_UPLOAD_BUCKET_URL which is https://pirateandfox-sops.s3.amazonaws.com

and

AWS_S3_UPLOAD_BUCKET_NAME which is pirateandfox-sops

So I'm thinking that outline then puts them both together into the request URL, and if that is done for everyone and works, I still don't understand why it wouldn't for me?

**Reply by ersel** [reply]
what happens if you set `AWS_S3_UPLOAD_BUCKET_URL` to `https://s3-<aws-region>.amazonaws.com` type URL along with the 
`AWS_S3_FORCE_PATH_STYLE=0`?

**Reply by ersel** [reply]
```
        AWS_S3_UPLOAD_BUCKET_URL: https://s3.us-east-2.amazonaws.com
        AWS_S3_UPLOAD_BUCKET_NAME: my-bucket-name
        AWS_S3_FORCE_PATH_STYLE: 0
```

setting up environment variables like above seems to work for me 👍

**Reply by justinhandley** [reply]
Thanks @ersel - removing the bucker from the bucket URL did the trick!

**Reply by tommoor** [reply]
I think clearing out `AWS_S3_UPLOAD_BUCKET_NAME=""` and using the original url should work too – the key is not to have it in both places. It's worth noting that AWS is deprecating not using the bucket name in the url.

I'm setting up a new installation in the next few days so I'll give it a try, sorry this isn't well documented yet.

**Reply by justinhandley** [reply]
> I think clearing out `AWS_S3_UPLOAD_BUCKET_NAME=""` and using the original url should work too – the key is not to have it in both places. It's worth noting that AWS is deprecating not using the bucket name in the url.
> 
> I'm setting up a new installation in the next few days so I'll give it a try, sorry this isn't well documented yet.

Actually, when I clear the bucket name, the server crashes with 

`Jan 23 18:11:55 pf-sops heroku/web.1 State changed from starting to crashed
Jan 23 18:11:55 pf-sops app/web.1 The AWS_S3_UPLOAD_BUCKET_NAME env variable must be set when using AWS`

**Reply by justinhandley** [reply]
@tommoor - so, I spoke too soon.  I don't know if these things are handled differently, but now I have image upload working on documents, however, when I try to upload either my profile image or the company logo, those are failing with a 403 error.  It is no longer the 405 error.  Is there a difference in the way files are uploaded for the profile images vs the document images?

**Reply by tommoor** [reply]
I'm setting up that new instance tonight, I'll put together docs once I get it all working :)

**Reply by tommoor** [reply]
Based on your description I think you're missing the `s3:PutObjectAcl` IAM policy, I'm putting together an example config here:
https://wiki.generaloutline.com/share/125de1cc-9ff6-424b-8415-0d58c809a40f

And there's a small supporting PR here for these new AWS bucket changes:
https://github.com/outline/outline/pull/1847

**Reply by justinhandley** [reply]
Hey @tommoor -  so, I added that (I actually had an approve * policy), so I changed to your example setup, and I also deployed the dev branch with that new change, which did mean that now I can put the full URL in my .env and have uploads work on the documents, so now, the only thing still not working, even adding the PutObjectAcl is the upload of the company logo and profile photos.

**Reply by tommoor** [reply]
These images are uploaded with `public` ACL, so they can be viewed on shared documents and other logged out screens. What do you have for "Block public access" under "Permissions"? It should be set to `off` so that the permissions can be set on a per object basis.

![image](https://user-images.githubusercontent.com/380914/106233245-e6d53880-61aa-11eb-86ea-94a6bb5589b9.png)

**Reply by justinhandley** [reply]
Thanks for your patience with this @tommoor  - that was the one.  Now fully functional across the board!!!

**Reply by tommoor** [reply]
Thanks for bearing with it, I added that detail to the doc too
